### PR TITLE
[SPARK-11246] [SQL] Table cache for Parquet broken in 1.5

### PR DESCRIPTION
**Pull request by xwu0226**
The root cause is that when spark.sql.hive.convertMetastoreParquet=true by default, the cached InMemoryRelation of the ParquetRelation can not be looked up from the cachedData of CacheManager because the key comparison fails even though it is the same LogicalPlan representing the Subquery that wraps the ParquetRelation.  
The solution in this PR is overriding the LogicalPlan.sameResult function in Subquery case class to eliminate subquery node first before directly comparing the child (ParquetRelation), which will find the key  to the cached InMemoryRelation.

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

**Comment by yhuai** [comment]
Actually, `sameResult` will remove `SubQuery` nodes. I think the problem is that `cleanRight.cleanArgs == cleanLeft.cleanArgs` returns false when two `LogicalRelation`s are considered generating the same result (because of `expectedOutputAttributes` that we added in #8889 ). I think the right fix is to correctly override the `cleanArgs` field of `LogicalRelation` to exclude `expectedOutputAttributes`.

**Comment by xwu0226** [comment]
@yhuai Thanks so much for your comments and suggestions!
I have thought through your suggestions and override `cleanArgs` field in `LogicalRelation`, that returns the `BaseRelation` (in this case, `ParquetRelation`) without the `expectedOutputAttributes`..  I removed the original fix from `Subquery` and commited the change. 

Let me know what you think. Thank you again!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 1]
```diff
@@ -61,6 +64,7 @@ case class LogicalRelation(
     case LogicalRelation(otherRelation, _) => relation == otherRelation
     case _ => false
   }
+  override lazy val cleanArgs: Seq[Any] = Seq(relation)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: Add a new line above this line.

**Comment by yhuai** [comment]
Just a minor comment about the format.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 2]
```diff
@@ -17,10 +17,13 @@
 package org.apache.spark.sql.execution.datasources
 
 import org.apache.spark.sql.catalyst.analysis.MultiInstanceRelation
-import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeMap, AttributeReference}
+import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan, Statistics}
+import org.apache.spark.sql.catalyst.trees.TreeNode
 import org.apache.spark.sql.sources.BaseRelation
 
+import scala.collection.mutable.ArrayBuffer
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: Import ordering (see https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala (hunk 1 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: Thanks! Will do.

**Comment by xwu0226** [comment]
@yhuai Just pushed again and also removed some unnecessary imports that I used for debugging today and forgot to remove.   
Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala [hunk 3]
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.sql.catalyst.plans.logical
 
+import org.apache.spark.sql.catalyst.analysis.EliminateSubQueries
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: also remove this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 4]
```diff
@@ -62,6 +62,8 @@ case class LogicalRelation(
     case _ => false
   }
 
+  override lazy val cleanArgs: Seq[Any] = Seq(relation)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: Can you also add a comment to explain why we need to override it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala (hunk 3 above)
Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Yes. I will. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala (hunk 4 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: Yes. let me add it. Thanks!

**Comment by xwu0226** [comment]
@yhuai Just pushed. Please help take a look, thanks!

**Comment by yhuai** [comment]
LGTM. Merging to master and branch 1.5.

**Comment by xwu0226** [comment]
@yhuai Thank you very much for merging it.

**Pull request closed without merging**